### PR TITLE
Fix SyntaxGenerator.InsertMembers for records

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/48789

image

### CODE REVIEW

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs [hunk 1]
```diff
@@ -1271,8 +1271,32 @@ private static int GetDeclarationCount(SyntaxNode declaration)
...
+
+                return recordDeclaration;
+            }
+
+            return declaration;
+        }
+
         public override SyntaxNode InsertMembers(SyntaxNode declaration, int index, IEnumerable<SyntaxNode> members)
         {
+            declaration = EnsureRecordDeclarationHasBody(declaration);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs:1299]: i'm not certain how this works with the code that exists below.  i.e. in the declaration switch lower on, we don't seem to handle records.

**Comment by Youssef1313** [comment]
If this goes through `RecordDeclarationSyntax.WithMembers` method, it might be better to update the auto-generated `Update` method IMO.

Current implementation is:

```csharp
        public RecordDeclarationSyntax Update(SyntaxList<AttributeListSyntax> attributeLists, SyntaxTokenList modifiers, SyntaxToken keyword, SyntaxToken identifier, TypeParameterListSyntax? typeParameterList, ParameterListSyntax? parameterList, BaseListSyntax? baseList, SyntaxList<TypeParameterConstraintClauseSyntax> constraintClauses, SyntaxToken openBraceToken, SyntaxList<MemberDeclarationSyntax> members, SyntaxToken closeBraceToken, SyntaxToken semicolonToken)
        {
            if (attributeLists != this.AttributeLists || modifiers != this.Modifiers || keyword != this.Keyword || identifier != this.Identifier || typeParameterList != this.TypeParameterList || parameterList != this.ParameterList || baseList != this.BaseList || constraintClauses != this.ConstraintClauses || openBraceToken != this.OpenBraceToken || members != this.Members || closeBraceToken != this.CloseBraceToken || semicolonToken != this.SemicolonToken)
            {
                var newNode = SyntaxFactory.RecordDeclaration(attributeLists, modifiers, keyword, identifier, typeParameterList, parameterList, baseList, constraintClauses, openBraceToken, members, closeBraceToken, semicolonToken);
                var annotations = GetAnnotations();
                return annotations?.Length > 0 ? newNode.WithAnnotations(annotations) : newNode;
            }
 
            return this;
        }
```

Instead of passing the braces and semicolon tokens as-is, it should check the count of members and remove the semicolon when appropriate.

**Comment by CyrusNajmabadi** [comment]
The underlying syntax apis very specifically do not do any processing of their inputs or contain any logic. That way all consumers can now exactly how 100% of that api will behave. It's important we keep that

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs (hunk 1 above)
Comment (jcouv) [reply src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs:1299]: records fall into `TypeDeclarationSyntax` case :-)

**Comment by jcouv** [comment]
@CyrusNajmabadi Did I answer your question satisfactorily?

**Comment by CyrusNajmabadi** [comment]
Yes.  I got confuesd between that type and BaseTypeDeclarationSyntax

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs [hunk 2]
```diff
@@ -1271,8 +1271,32 @@ private static int GetDeclarationCount(SyntaxNode declaration)
                 _ => 1,
             };
 
+        private static SyntaxNode EnsureRecordDeclarationHasBody(SyntaxNode declaration)
+        {
+            if (declaration is RecordDeclarationSyntax recordDeclaration)
+            {
+                if (recordDeclaration.SemicolonToken.IsKind(SyntaxKind.SemicolonToken))
+                {
+                    recordDeclaration = recordDeclaration.WithSemicolonToken(default);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs:1280]: just do htis unilaterally.

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs [hunk 3]
```diff
@@ -1271,8 +1271,32 @@ private static int GetDeclarationCount(SyntaxNode declaration)
...
+                {
+                    recordDeclaration = recordDeclaration.WithOpenBraceToken(SyntaxFactory.Token(SyntaxKind.OpenBraceToken));
+                }
+                if (!recordDeclaration.CloseBraceToken.IsKind(SyntaxKind.CloseBraceToken))
+                {
+                    recordDeclaration = recordDeclaration.WithCloseBraceToken(SyntaxFactory.Token(SyntaxKind.CloseBraceToken));
+                }
+
+                return recordDeclaration;
+            }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs:1292]:
```suggestion
                return recordDeclaration.WithSemicolonToken(default)
                    .WithOpenBraceToken(recordDeclaration.OpenBraceToken == default ? Token(SyntaxKind.OpenBraceToken) : recordDeclaration.OpenBraceToken)
                    .WithCloseBraceToken(recordDeclaration.CloseBraceToken == default ? Token(SyntaxKind.CloseBraceToken) : recordDeclaration.CloseBraceToken);
            }
```

File: src/Workspaces/CSharpTest/CodeGeneration/SyntaxGeneratorTests.cs [hunk 4]
```diff
@@ -2991,6 +2991,60 @@ public void TestMultiFieldDeclarations()
 }");
         }
 
+        [Fact]
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharpTest/CodeGeneration/SyntaxGeneratorTests.cs:2994]: PITA.  but please add WorkItem here and for the other otests.

**Review by CyrusNajmabadi — APPROVED** [review]

**Pull request merged**